CIP-????/README.md

```diff
@@ -0,0 +1,97 @@
+
+#### Order matching
+
+A common pattern in DEXs is to have a list of inputs/outputs to match up in a datum. 
+For example, we might have:
+```
+inputIdxs :: BuiltinList Integer
+outputIdxs :: BuiltinList Integer
+```
+We then want to go through these lists, looking up the corresponding inputs and outputs and check some property (e.g. that the value is directly transferred from one to the other).
+
+This requires a quadratic amount of work, which puts a low ceiling on how many orders can be processed at once. 
+Empirically, many are capped at about 30, whereas if they were limited only by the amount of space in the transaction for inputs and outputs the limit would be hundreds.
+
+If we had arrays with constant time indexing, we could make this linear instead (we would still need to do a linear amount of work to create arrays for the transaction inputs and outputs from the lists in the script context).
+
+#### `Constr` arguments
+
+The `Data` type has a `Constr` alternative which is used for encoding datatype constructors. 
+This is used for encoding the script context, and is used by languages such as Aiken extensively for representing user-defined datatypes also.
+
+The fields of the constructor are encoded in a list; hence to access a particular field the compiled code needs to do a linear amount of work. 
+If the arguments to a `Constr` were an array, we could access the fields in constant time.
+
+### Maps
+
+#### Operations on `Value`
+
+The `Value` type is a nested map: it is a map from bytestrings (representing policy IDs) to maps from bytestrings (representing token names) to integers (representing quantities).
+Since map operations are currently linear, this means that even simple operations like checking whether one value is less than another can have quadratic cost.
+
+This would be much better if map operations were logarithmic cost.
+
+#### Indexing by party
+
+Many applications have a known set of participants identified by some bytestring, typically a public key. 
+It is therefore natural to store per-party state in a map indexed by the party identifier.
+
+Since map operations are currently linear, this is needlessly expensive and imposes a limit on the number of parties.
+
+## Goals
+
+1. Reduce the cost of operations on `Value` by a factor of 2-10
+2. Reduce the cost of a matching algorithm such that we can handle hundreds of matches for the same cost it currently takes to do 30.
+
+## Open questions
+
+- Can we implement a set/map data structure in Plutus Core code that has acceptable performance and doesn’t require too much size overhead?
+- Do we need generic maps or is a map-from-bytestring sufficient? What about map-from-integer?
+    - Generic maps are harder since we typically need to know how to order the key type
+- Is an array type useful even if it is immutable?
+    - We are unlikely to be able to offer mutable arrays
+- Are builtin data structures useful enough even if they can only contain builtin types?
```

Comment:
`Value` alone should make this useful. Also Aiken users would certainly answer "yes".

Reply:
Yes, I think those two are compelling